universum/modules/output/html_output.py

```diff
@@ -15,68 +15,57 @@ def __init__(self, *args, **kwargs):
         self._filename = None
         self.artifact_dir_ready = False
         self._log_buffer = list()
+        self._block_level = 0
 
     def set_artifact_dir(self, artifact_dir):
         self._filename = os.path.join(artifact_dir, "log.html")
 
     def open_block(self, num_str, name):
-        self.log(name) # stub
+        self._log_line(f"{num_str} {name}")
+        self._block_level += 1
 
     def close_block(self, num_str, name, status):
-        self.log(name) # stub
+        self._block_level -= 1
+        indent = "  " * self._block_level
+        self._log_line(f"{indent} \u2514 [{status}]")
+        self._log_line("")
 
     def report_error(self, description):
-        self.log(description) # stub
+        pass
 
     def report_skipped(self, message):
-        self.log(message) # stub
+        self._log_line(message)
 
     def report_step(self, message, status):
-        self.log(message) # stub
+        self.log(message)
 
     def change_status(self, message):
-        self.log(message) # stub
+        pass
 
     def log_exception(self, line):
-        self.log(line) # stub
+        self._log_line(f"Error: {line}")
 
     def log_stderr(self, line):
-        self.log(line) # stub
+        self._log_line(f"stderr: {line}")
 
     def log(self, line):
-        self._log_line(line) # stub
+        self._log_line(f"==> {line}")
```

Comment:
Isn't this behaviour replicating the `TerminalBasedOutput`? Shouldn't it be integrated in some way?

Reply:
Yes, common logic will be extracted from `TerminalBasedOutput` and `HtmlOutput` and moved to some parent class. But it will be done after HTML output implementation to avoid double-work.